arangod/Graph/Enumerators/YenEnumerator.cpp

```diff
@@ -0,0 +1,380 @@
+  _resourceMonitor.decreaseMemoryUsage(_totalMemoryUsageHere);
+}
+
+template<class ProviderType, class EnumeratorType, bool IsWeighted>
+auto YenEnumerator<ProviderType, EnumeratorType, IsWeighted>::destroyEngines()
+    -> void {
+  _shortestPathEnumerator->destroyEngines();
+}
+
+template<class ProviderType, class EnumeratorType, bool IsWeighted>
+void YenEnumerator<ProviderType, EnumeratorType, IsWeighted>::clear() {
+  _shortestPathEnumerator->clear();
+  _shortestPaths.clear();
+  _candidatePaths.clear();
+  _arena.clear();
+  _isDone = true;
+  _isInitialized = false;
+}
+
+/**
+ * @brief Quick test if the finder can prove there is no more data available.
+ *        It can respond with false, even though there is no path left.
+ * @return true There will be no further path.
+ * @return false There is a chance that there is more data available.
+ */
+template<class ProviderType, class EnumeratorType, bool IsWeighted>
+bool YenEnumerator<ProviderType, EnumeratorType, IsWeighted>::isDone() const {
+  // This is more subtle than first meets the eye: If we are not yet
+  // initialized, we must return `true`.
+  // This is necessary such that the EnumeratePathsExecutor works. Once
+  // we are initialized with `reset()`, we return `true` once we have
+  // proved that no further path will be found. Note that it might be that
+  // we have returned `false` and yet no further path is found.
+  return _isDone;
+}
+
+/**
+ * @brief Reset to new source and target vertices.
+ * This API uses string references, this class will not take responsibility
+ * for the referenced data. It is caller's responsibility to retain the
+ * underlying data and make sure the strings stay valid until next
+ * call of reset.
+ *
+ * @param source The source vertex to start the paths
+ * @param target The target vertex to end the paths
+ */
+template<class ProviderType, class EnumeratorType, bool IsWeighted>
+void YenEnumerator<ProviderType, EnumeratorType, IsWeighted>::reset(
+    VertexRef source, VertexRef target, size_t depth) {
+  _source = source;
+  _target = target;
+  clear();
+  _isDone = false;
+  _isInitialized = true;
+}
+
+template<class ProviderType, class EnumeratorType, bool IsWeighted>
+PathResult<ProviderType, typename ProviderType::Step>
+YenEnumerator<ProviderType, EnumeratorType, IsWeighted>::toOwned(
+    PathResult<ProviderType, typename ProviderType::Step> const& path) {
+  PathResult<ProviderType, typename ProviderType::Step> copy{
+      path.getSourceProvider(), path.getTargetProvider()};  // empty path
+  size_t l = path.getLength();
+  for (size_t i = 0; i < l + 1; ++i) {
+    typename ProviderType::Step::Vertex v{
+        _arena.toOwned(path.getVertex(i).getID())};
+    copy.appendVertex(v);
+    if (i == l) {
+      break;  // one more vertex than edge
+    }
+    typename ProviderType::Step::Edge e{
+        _arena.toOwned(path.getEdge(i).getID())};
+    copy.appendEdge(e, path.getWeight(i));
+  }
```

Comment:
I am not sure I understand the questions. When we don't have the Taboo wrapper, then no `GraphArena` is ever used. This is because the `YenEnumerator` needs the taboo wrapper and the `GraphArena` is only ever used in the `YenEnumerator`.
The problem here is that the whole infrastructure around `TwoSidedEnumerator` and `WeightedTwoSidedEnumerator` which we are using here only works with references for vertex IDs and edge IDs and not with owned types. We have to adjust to this, since we are using this. And since we have to `reset` the sub-enumerators for every shortest path search all references we got become invalid.

As far as I see, since we get references out of the algorithms we use and these references become invalid before we are done with the results, we have to copy somewhere. `GraphArena` is the simplest thing I could do, as far as I see.

But happy to discuss further.

Reply:
To answer your questions: In traversals we usually get out edges and vertices only once. When enumerating all paths, it is entirely possible to get out vertices and edges multiple times. For Yen specifically, I expect that users will not get out extremely many paths. So it is almost certainly not worthwhile the effort (in complexity and performance) to deduplicate the vertex and edge ids in the paths, just to save some memory.

Reply:
Aha ok thanks for claryfing this, I will take another look to see I think I missed the part that this is only used with TabooWrapper